terrarium/bi_ci/bi_ci/split_pytest_tasks.py

```diff
@@ -69,101 +84,204 @@ def print_error(message: str) -> None:
 
-def read_pytest_targets(path: pathlib.Path) -> list[PyprojectPytestTarget]:
-    if not path.is_file():
-        print_error(f"File {path.absolute()} not found")
-        raise FileNotFoundError(f"File {path.absolute()} not found")
+def read_pytest_targets(absolute_path: pathlib.Path, relative_path: str) -> list[PyprojectPytestTarget]:
+    pyproject_path = absolute_path / "pyproject.toml"
+
+    if not pyproject_path.is_file():
+        error_message = f"pyproject.toml not found in {relative_path}"
+        print_error(error_message)
+        raise FileNotFoundError(error_message)
 
-    with open(path, "r") as file:
+    with open(pyproject_path, "r") as file:
         toml_data = tomlkit.parse(file.read())
 
     raw_targets = toml_data.get("datalens", {}).get("pytest", {})
 
     if len(raw_targets) == 0:
-        return [PyprojectPytestTarget()]
+        return [PyprojectPytestTarget(root_dir=f"{absolute_path.name}_tests/")]
 
-    return [PyprojectPytestTarget.from_dict(name, data) for name, data in raw_targets.items()]
+    return [PyprojectPytestTarget.from_raw(name=name, data=data) for name, data in raw_targets.items()]
 
 
 def get_target_tests(
-    package_path: str,
-    requested_modes: set[str],
+    relative_path: str,
+    requested_labels: set[str],
     target: PyprojectPytestTarget,
     raise_on_unused_label: bool = False,
 ) -> typing.Generator[TestTarget, None, None]:
-    requested_labels = requested_modes & target.labels
+    requested_labels = requested_labels & target.labels
+    unused_labels = target.labels - requested_labels
 
-    if unused_labels := target.labels - requested_labels:
-        error_message = f"Unused labels in {package_path}:{target.name}: {' '.join(unused_labels)}"
+    if len(unused_labels) > 0:
+        error_message = f"Unused labels in {relative_path}:{target.name}: {' '.join(unused_labels)}"
         print_error(error_message)
 
         if raise_on_unused_label:
             raise ValueError(error_message)
 
     for requested_mode in requested_labels:
-        yield TestTarget(package_path=package_path, section=target.name, requested_mode=requested_mode)
+        yield TestTarget(relative_path=relative_path, section=target.name, requested_mode=requested_mode)
+
+
+def is_parent_path(parent: pathlib.Path, child: pathlib.Path) -> bool:
+    return str(child).startswith(str(parent))
+
+
+def get_tests_root_dirs(
+    package_path: pathlib.Path,
+    pytest_targets: list[PyprojectPytestTarget],
+) -> set[pathlib.Path]:
+    all_root_dirs: set[pathlib.Path] = set()
+
+    # adding default test diirectory
+    all_root_dirs.add(package_path / f"{package_path.name}_tests")
+
+    for target in pytest_targets:
+        root_dir = package_path / target.root_dir
+        all_root_dirs.add(root_dir)
+
+        if not root_dir.is_dir():
+            print_error(f"Root dir {root_dir.name} not found in {root_dir.parent}")
+
+    # deduplicating subdirectories
+
+    root_dirs: set[pathlib.Path] = set()
```

Comment:
can be a oneliner:
`root_dirs = {root_dir for root_dir, other_root_dir in itertools.product(all_root_dits, repeat=2) if is_parent_path(parent=other_root_dir, child=root_dir) and other_root_dir != root_dir}`

or just use `itertools.product` instead of nested loops